src/services/statistics.service.ts

```diff
@@ -0,0 +1,161 @@
+import {
+  Injectable,
+  BadRequestException,
+  InternalServerErrorException,
+} from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository, SelectQueryBuilder } from 'typeorm';
+import { Appointment } from '../entities/index';
+import { ResponseStatisticsDto } from '../dtos';
+import { plainToClass } from 'class-transformer';
+
+@Injectable()
+export default class StatisticsService {
+  constructor(
+    @InjectRepository(Appointment)
+    private readonly appointmentRepository: Repository<Appointment>,
+  ) {}
+
+  async getStatistics(startDate: string, endDate: string, treatment?: string) {
+    if (!startDate || !endDate) {
+      throw new BadRequestException(
+        'Las fechas de inicio y fin son obligatorias.',
+      );
+    }
+
+    const [
+      totalEarnings,
+      totalServices,
+      totalCompletedAppointments,
+      totalCanceledAppointments,
+    ] = await Promise.all([
+      this.getTotalEarnings(startDate, endDate, treatment),
+      this.getTotalServices(startDate, endDate, treatment),
+      this.getTotalCompletedAppointments(startDate, endDate, treatment),
+      this.getTotalCanceledAppointments(startDate, endDate, treatment),
+    ]);
+
+    return plainToClass(ResponseStatisticsDto, {
+      totalEarnings,
+      totalServices,
+      totalCompletedAppointments,
+      totalCanceledAppointments,
+    });
+  }
+
+  private applyFilters(
+    qb: SelectQueryBuilder<Appointment>,
+    startDate: string,
+    endDate: string,
+    treatment?: string,
+  ): SelectQueryBuilder<Appointment> {
+    qb.where(
+      'DATE(appointment.scheduled_start) BETWEEN :startDate AND :endDate',
+      { startDate, endDate },
+    );
+
+    if (treatment) {
+      qb.innerJoin(
+        'appointment.treatments',
+        'treatment',
+        'treatment.name = :treatment',
+        { treatment },
+      );
+    }
+
+    return qb;
+  }
+
+  private async getTotalEarnings(
+    startDate: string,
+    endDate: string,
+    treatment?: string,
+  ): Promise<number> {
+    try {
+      const qb = this.appointmentRepository
+        .createQueryBuilder('appointment')
+        .select('COALESCE(SUM(appointment.total_price), 0)', 'totalearnings');
+
+      this.applyFilters(qb, startDate, endDate, treatment);
```

Comment:
shouldn't you reassign the variable here to store the query concatenating the treatment filters?:

```suggestion
         let qb = this.appointmentRepository
          .createQueryBuilder('appointment')
          .select('COALESCE(SUM(appointment.total_price), 0)', 'totalearnings');
          
        let qb = this.applyFilters(qb, startDate, endDate, treatment);
```

This is a question

Reply:
The applyFilters method accepts a QueryBuilder object. Objects in JavaScript/TypeScript are passed by reference, not by value

Reply:
yeah, that's true. But my point is more about the immutability, so you're saying that when sending the QueryBuilder instance you’re mutating the QueryBuilder directly.

Got it, tricky, but it's okay then